core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java

```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
 
   static Metrics fromWriter(DatumWriter<?> datumWriter, Schema schema, long numRecords,
                             MetricsConfig inputMetricsConfig) {
-    // TODO will populate in following PRs if datum writer is a MetricsAwareDatumWriter
-    return new Metrics(numRecords, null, null, null);
+    if (!(datumWriter instanceof MetricsAwareDatumWriter)) {
+      return new Metrics(numRecords, null, null, null, null);
+    }
+
+    MetricsAwareDatumWriter<?> metricsAwareDatumWriter = (MetricsAwareDatumWriter<?>) datumWriter;
+    MetricsConfig metricsConfig;
+    if (inputMetricsConfig == null) {
+      metricsConfig = MetricsConfig.getDefault();
+    } else {
+      metricsConfig = inputMetricsConfig;
+    }
+
+    Map<Integer, Long> valueCounts = new HashMap<>();
+    Map<Integer, Long> nullValueCounts = new HashMap<>();
+    Map<Integer, Long> nanValueCounts = new HashMap<>();
+    Map<Integer, ByteBuffer> lowerBounds = new HashMap<>();
+    Map<Integer, ByteBuffer> upperBounds = new HashMap<>();
+
+    metricsAwareDatumWriter.metrics().forEach(metrics -> {
+      String columnName = schema.findColumnName(metrics.id());
+      MetricsModes.MetricsMode metricsMode = metricsConfig.columnMode(columnName);
+      if (metricsMode == MetricsModes.None.get()) {
+        return;
+      }
+
+      valueCounts.put(metrics.id(), metrics.valueCount());
+      nullValueCounts.put(metrics.id(), metrics.nullValueCount());
+      Type type = schema.findType(metrics.id());
+
+      if (type.typeId() == Type.TypeID.FLOAT || type.typeId() == Type.TypeID.DOUBLE) {
+        nanValueCounts.put(metrics.id(), metrics.nanValueCount());
+      }
+
+      if (metricsMode == MetricsModes.Counts.get()) {
+        return;
+      }
+
+      updateLowerBound(metrics, type, metricsMode).ifPresent(lowerBound -> lowerBounds.put(metrics.id(), lowerBound));
+      updateUpperBound(metrics, type, metricsMode).ifPresent(upperBound -> upperBounds.put(metrics.id(), upperBound));
```

Comment:
I don't quite understand the decision to return an option instead of just passing `lowerBounds` or `upperBounds` into the update method and having the `put` happen there. Wouldn't it be simpler if the put happened at the end of `updateLowerBound` or `updateUpperBound`?